spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java

```diff
@@ -0,0 +1,319 @@
+    if (!DataTypes.IntegerType.acceptsType(dt) && !DataTypes.LongType.acceptsType(dt) &&
+        !DataTypes.StringType.acceptsType(dt) && !DataTypes.BinaryType.acceptsType(dt) &&
+        !(dt instanceof DecimalType)) {
+      throw new UnsupportedOperationException(
+          "Cannot truncate input type. Expected one of [IntegerType, LongType, DecimalType, StringType, BinaryType], " +
+              "but found " + dt.catalogString());
+    }
+  }
+
+  private static void validateTruncationWidthType(DataType widthType) {
+    if (!(widthType instanceof IntegerType ||
+        !(widthType instanceof LongType && DataTypes.IntegerType.acceptsType(widthType)))) {
+      throw new UnsupportedOperationException("Expected truncation width to be an integer type");
+    }
+  }
+
+  @Override
+  public BoundFunction bind(StructType inputType) {
+    if (inputType.fields().length != 2) {
+      throw new UnsupportedOperationException(
+          String.format("Invalid input type. Expected 2 fields but found %s", inputType.fields().length));
+    }
+
+    StructField fieldToTruncate = inputType.apply(0);
+    StructField truncationWidth = inputType.apply(1);
+
+    DataType toTruncateSparkType = fieldToTruncate.dataType();
+    Type sourceIcebergType = SparkSchemaUtil.convert(fieldToTruncate.dataType());
+
+    validateSourceTypeUsingSparkTypes(fieldToTruncate.dataType());
+    validateTruncationWidthType(truncationWidth.dataType());
+
+    switch (sourceIcebergType.typeId()) {
+      case INTEGER:
+        return new TruncateInteger();
+      case LONG:
+        return new TruncateLong();
+      case DECIMAL:
+        return new TruncateDecimal(
+            ((DecimalType) toTruncateSparkType).precision(),
+            ((DecimalType) toTruncateSparkType).scale());
+      case STRING:
+        return new TruncateString();
+      case BINARY:
+        return new TruncateBinary();
+      default:
+        throw new UnsupportedOperationException(
+            "Cannot truncate type: " + sourceIcebergType.typeId());
+    }
+  }
+
+  @Override
+  public String description() {
+    return "Truncate - The Iceberg truncate function used for truncate partition transformations.";
+  }
+
+  @Override
+  public String name() {
+    return "truncate";
+  }
+
+  public abstract static class TruncateBase<T> implements ScalarFunction<T> {
+    @Override
+    public String name() {
+      return "truncate";
+    }
+
+    @Override
+    public boolean isResultNullable() {
+      return true;
+    }
+
+    @Override
+    public boolean isDeterministic() {
+      return true;
+    }
+  }
+
+  public static class TruncateInteger extends TruncateBase<Integer> {
+    // magic method used in codegen
+    public static int invoke(int value, int width) {
+      if (width <= 0) {
+        throw new IllegalArgumentException(String.format("Invalid truncate width: %s (must be > 0)", width));
+      }
+
+      return value - (((value % width) + width) % width);
```

Comment:
I'd prefer to use the actual truncate function. I talked with Dan and heard that the motivation for this was that width might change and we don't want to create a new instance of the transform every time. I think it's good to catch that possibility, but there are other options that avoid copying the transform implementation.

The option I like best is to assume that the width is not going to change across calls and check that width is the same as the last row's width:

```java
  private int lastWidth = -1;
  private Transform<Integer, Integer> transform = null;
  public static int invoke(int width, int value) {
    ...

    if (lastWidth != width) {
      this.lastWidth = width;
      this.transform = Transforms.truncate(IntegerType.get(), width);
    }

    return transform.apply(value);
  }
```

For constant values, this will work great. There's some overhead for changing widths, but I don't think that's going to be a common case since it makes very little sense.

Reply:
Agreed that it's not likely to be a common case. That's not really the intended purpose of the function and I don't care for the code duplication.

Will go with that approach then. Thank you!

Reply:
Actually, I'm reconsidering this. The option of having a static function that performs this would allow us to avoid boxing/unboxing and the lookup entirely. That's probably the better thing to do.

Reply:
I moved the truncation functions into `TruncateUtil` in api.

Some things already had utility functions handling their truncation, such as `UnicodeUtil.truncateString` so I didn't add functions for that.

The code in `TruncateUtil` explicitly assumes that the input has been validated, to allow for more efficient codegen or for skipping validation if it's already been done.
We pre-validate:
1. The Truncate `Transform` validates the `width` in their constructor so it doesn't need to be done again.
2. The Spark Truncate `BoundFunction` will immediately return `null` for `null` input for primitives.

Not doing duplicate validation was chosen because this code is in the hot path (namely because it gets included in Spark's codegen when used with `FunctionCatalog`).

Reply:
Extracted out a `TruncateUtil`, at least for the calls to be shared between `Truncate` transform and the generated code. As both `invoke` and the `TruncateUtil` are both static functions, will the Spark generated code remove the nested static function call?